velox/functions/lib/Re2Functions.cpp

```diff
@@ -1040,56 +1201,139 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> re2ExtractSignatures() {
   };
 }
 
-std::string unescape(
-    std::string_view pattern,
+namespace {
+// Sub-pattern's stats in a top-level pattern.
+struct SubPatternStats {
+  SubPatternKind kind;
+  // Count of this patterns in total.
+  size_t count = 0;
+  // First index of this pattern kind.
+  int32_t firstIndex = -1;
+  // Last index of this pattern kind.
+  int32_t lastIndex = -1;
+
+  void update(size_t index) {
+    count++;
+    if (firstIndex == -1) {
+      firstIndex = index;
+    }
+
+    lastIndex = index;
+  }
+};
+} // namespace
+
+// Return the length of the fixed part of the pattern.
+// Mainly used for kAtLeastN pattern.
+size_t fixedLength(
+    const std::vector<SubPatternKind>& subPatternKinds,
+    const std::vector<std::pair<size_t, size_t>>& subPatternRanges) {
+  size_t result = 0;
+  for (auto i = 0; i < subPatternKinds.size(); i++) {
+    if (subPatternKinds[i] != SubPatternKind::kAnyCharsWildcard) {
+      result += subPatternRanges[i].second;
```

Comment:
It is a bit confusing that we are summing up bytes with characters here. For example, we add number of characters for kSingleCharWildcard pattern and number of bytes for kLiteralString. Is this intentional? How does this work?

Reply:
`fixedLength`  is used to get the length of the fixed part(literal chars or single char wildcard) of a pattern:

* if the pattern is `kFixed`, then fixedLength = length of the pattern = length of the matched input
* if the pattern is `kRelaxedFixed` then fixedLength = length of the pattern **!=** length of the matched input